sdks/python/apache_beam/runners/portability/fn_api_runner/watermark_manager.py

```diff
@@ -0,0 +1,171 @@
+      self.inputs: Set[WatermarkManager.PCollectionNode] = set()
+      self.side_inputs: Set[WatermarkManager.PCollectionNode] = set()
+      self.outputs: Set[WatermarkManager.PCollectionNode] = set()
+
+    def __str__(self):
+      return 'StageNode<inputs=%s,side_inputs=%s' % (
+          [i.name for i in self.inputs], [i.name for i in self.side_inputs])
+
+    def output_watermark(self):
+      if not self.outputs:
+        return self.input_watermark()
+      else:
+        return min(o.watermark() for o in self.outputs)
+
+    def input_watermark(self):
+      if not self.inputs:
+        return timestamp.MAX_TIMESTAMP
+      w = min(i.upstream_watermark() for i in self.inputs)
+
+      if self.side_inputs:
+        w = min(w, min(i.upstream_watermark() for i in self.side_inputs))
+      return w
+
+  def __init__(self, stages):
+    # type: (List[translations.Stage]) -> None
+    self._pcollections_by_name: Dict[Union[str, translations.TimerFamilyId],
+                                     WatermarkManager.PCollectionNode] = {}
+    self._stages_by_name: Dict[str, WatermarkManager.StageNode] = {}
+
+    for s in stages:
+      stage_name = s.name
+      stage_node = WatermarkManager.StageNode(stage_name)
+      self._stages_by_name[stage_name] = stage_node
+
+      def add_pcollection(
+          pcname: str, snode: WatermarkManager.StageNode
+      ) -> WatermarkManager.PCollectionNode:
+        if pcname not in self._pcollections_by_name:
+          self._pcollections_by_name[pcname] = WatermarkManager.PCollectionNode(
+              pcname)
+        pcnode = self._pcollections_by_name[pcname]
+        assert isinstance(pcnode, WatermarkManager.PCollectionNode)
+        snode.inputs.add(pcnode)
+        return pcnode
+
+      # 1. Get stage inputs, create nodes for them, add to _stages_by_name,
+      #    and add as inputs to stage node.
+      for transform in s.transforms:
+        if transform.spec.urn == bundle_processor.DATA_INPUT_URN:
+          buffer_id = transform.spec.payload
+          if buffer_id == translations.IMPULSE_BUFFER:
+            pcoll_name = transform.unique_name
+            add_pcollection(pcoll_name, stage_node)
+            continue
+          else:
+            _, pcoll_name = split_buffer_id(buffer_id)
+          add_pcollection(pcoll_name, stage_node)
+
+      # 2. Get stage timers, and add them as inputs to the stage.
+      for transform in s.transforms:
+        if transform.spec.urn in translations.PAR_DO_URNS:
+          payload = proto_utils.parse_Bytes(
+              transform.spec.payload, beam_runner_api_pb2.ParDoPayload)
+          for timer_family_id in payload.timer_family_specs.keys():
+            timer_pcoll_name = (transform.unique_name, timer_family_id)
+            self._pcollections_by_name[
+                timer_pcoll_name] = WatermarkManager.PCollectionNode(
+                    timer_pcoll_name)
+            timer_pcoll_node = self._pcollections_by_name[timer_pcoll_name]
```

Comment:
why do we need these assertions immediately after updating the map?

Reply:
these are added to fix type checking issues (ensuring the types are what's expected)